datumaro/components/format_detection.py

```diff
@@ -0,0 +1,152 @@
+# Copyright (C) 2021 Intel Corporation
+#
+# SPDX-License-Identifier: MIT
+
+from enum import IntEnum
+from typing import Callable, Optional, Sequence
+import glob
+import os.path as osp
+
+
+class FormatDetectionConfidence(IntEnum):
+    """
+    Represents the level of confidence that a detector has in a dataset
+    belonging to the detector's format.
+    """
+
+    # Note to developers: more confidence levels could be added in the future,
+    # but they should all have positive values. This is to ensure that 0
+    # can be used as a sentinel value for comparing confidence levels.
+    LOW = 10
+    """
+    The dataset seems to belong to the format, but the format is too loosely
+    defined to be able to distinguish it from other formats.
+    """
+    MEDIUM = 20
+    """
+    The dataset seems to belong to the format, and is likely not to belong
+    to any other format.
+    """
+    # There's no HIGH confidence yet, because none of the detectors
+    # deserve it. It's reserved for when the detector is sure that
+    # the dataset belongs to the format; for example, because the format
+    # has explicit identification via magic numbers/files.
+
+class FormatRequirementsUnmet(Exception):
```

Comment:
> How would it be a tuple? It would just be a value of one of the supported types (IOW, a typing.Union).

I propose the following design:

```python
def apply_format_detector(
    dataset_root_path: str, detector: FormatDetector,
) -> Tuple[FormatDetectionConfidence, List[FailedFormatDetection]]:
    ... # do stuff
    if (success):
        return confidence, []
    else:
        return 0, errors_in_alternatives

def Environment.detect_format(
    dataset_root: str
) -> Tuple[List[str], Dict[str, List[FailedFormatDetection]]:
    max_conf = 0
    all_errors = {}
    found_formats = []

    for format in formats:
        conf, errors = appy_format_detector(dataset_root, format.detect)
        if conf > max_conf:
            max_conf = conf
            found_formats = []
            # maybe transfer existing results to the failed list with a LowPriorityError
        if conf > 0 and conf == max_conf:
            found_formats.append(format_name)
        else:
            all_errors[format_name] = errors
    return found_formats, all_errors
```

Reply:
What is the benefit of using a `Tuple` return type for `apply_format_detector` rather than `Union`? A `Union` would ensure at the type system level that a confidence level is mutually exclusive with an error list. It's also a more direct equivalent of `std::expected` from the C++ proposal.

Reply:
Forgot to add this...

> This way we can return all the accumulated information to the user.

I think this could ultimately be a good thing to do, but for this PR I'd like to focus on the detection backend and not change the interface of `Environment.detect_dataset`.

Reply:
> What is the benefit of using a Tuple return type for apply_format_detector rather than Union? A Union would ensure at the type system level that a confidence level is mutually exclusive with an error list. It's also a more direct equivalent of std::expected from the C++ proposal.

Probably, in Python, there is no big problem in checking the value type, except it can be difficult to be written precisely (eg. `-> Union[Dict[str, int], Dict[str, str]]`). I think, both ways are valid and working.